libs/config/yamlloader/loader.go

```diff
@@ -0,0 +1,227 @@
+}
+
+func newLoader(path string) *loader {
+	return &loader{
+		path: path,
+	}
+}
+
+func (d *loader) location(node *yaml.Node) config.Location {
+	return config.Location{
+		File:   d.path,
+		Line:   node.Line,
+		Column: node.Column,
+	}
+}
+
+func (d *loader) load(node *yaml.Node) (config.Value, error) {
+	loc := config.Location{
+		File:   d.path,
+		Line:   node.Line,
+		Column: node.Column,
+	}
+
+	var value config.Value
+	var err error
+
+	switch node.Kind {
+	case yaml.DocumentNode:
+		value, err = d.loadDocument(node, loc)
+	case yaml.SequenceNode:
+		value, err = d.loadSequence(node, loc)
+	case yaml.MappingNode:
+		value, err = d.loadMapping(node, loc)
+	case yaml.ScalarNode:
+		value, err = d.loadScalar(node, loc)
+	case yaml.AliasNode:
+		value, err = d.loadAlias(node, loc)
+	default:
+		return config.NilValue, errorf(loc, "unknown node kind: %v", node.Kind)
+	}
+
+	if err != nil {
+		return value, err
+	}
+
+	// Mark value as anchor if needed.
+	// If this node doesn't map to a type, we don't need to warn about it.
+	if node.Anchor != "" {
+		value = value.MarkAnchor()
+	}
+
+	return value, nil
+}
+
+func (d *loader) loadDocument(node *yaml.Node, loc config.Location) (config.Value, error) {
+	return d.load(node.Content[0])
+}
+
+func (d *loader) loadSequence(node *yaml.Node, loc config.Location) (config.Value, error) {
+	acc := make([]config.Value, len(node.Content))
+	for i, n := range node.Content {
+		v, err := d.load(n)
+		if err != nil {
+			return config.NilValue, err
+		}
+
+		acc[i] = v
+	}
+
+	return config.NewValue(acc, loc), nil
+}
+
+func (d *loader) loadMapping(node *yaml.Node, loc config.Location) (config.Value, error) {
+	var merge *yaml.Node
+
+	acc := make(map[string]config.Value)
+	for i := 0; i < len(node.Content); i += 2 {
+		key := node.Content[i]
+		val := node.Content[i+1]
+
+		// Assert that keys are strings
+		if key.Kind != yaml.ScalarNode {
+			return config.NilValue, errorf(loc, "key is not a scalar")
+		}
+
+		st := key.ShortTag()
+		switch st {
+		case "!!str":
+			// OK
+		case "!!merge":
+			merge = val
+			continue
+		default:
+			return config.NilValue, errorf(loc, "invalid key tag: %v", st)
+		}
+
+		v, err := d.load(val)
+		if err != nil {
+			return config.NilValue, err
+		}
+
+		acc[key.Value] = v
+	}
+
+	if merge == nil {
+		return config.NewValue(acc, loc), nil
+	}
+
+	// Build location for the merge node.
+	var mloc = d.location(merge)
+	var merr = errorf(mloc, "map merge requires map or sequence of maps as the value")
+
+	// Build a sequence of values to merge.
+	// The entries that we already accumulated have precedence.
+	var seq []map[string]config.Value
+	switch merge.Kind {
+	case yaml.SequenceNode:
+		for _, n := range merge.Content {
+			v, err := d.load(n)
+			if err != nil {
+				return config.NilValue, err
+			}
+			m, ok := v.AsMap()
+			if !ok {
+				return config.NilValue, merr
+			}
+			seq = append(seq, m)
+		}
+	case yaml.AliasNode:
+		v, err := d.load(merge)
```

Comment:
Do we ever intend this pathway and that inside the loop in the SequenceNode branch to diverge? If not, it may make sense to abstract this logic into a standalone function to ensure consistency over time.

Reply:
We do not. I rewrote this to first get a slice of `*yaml.Node` and then do the processing to abstract it.